pkg/sql/vecindex/vecstore/store_txn.go

```diff
@@ -309,10 +333,128 @@ func (tx *Txn) SearchPartitions(
 	return nil
 }
 
+// InitGetFullVectorsFetchSpec initializes the fetch spec for GetFullVectors.
+// It is used to fetch unquantized vectors from the primary index. The
+// 'sourceIndex' is the index from which to read the unquantized vectors and is
+// normally the primary key, but backfill will sometimes want to specify a
+// different index.
+func InitGetFullVectorsFetchSpec(
+	spec *vecstorepb.GetFullVectorsFetchSpec,
+	evalCtx *eval.Context,
+	tableDesc catalog.TableDescriptor,
+	indexDesc catalog.Index,
+	sourceIndex catalog.Index,
+) error {
+	vectorColID := indexDesc.VectorColumnID()
+	if err := rowenc.InitIndexFetchSpec(
+		&spec.FetchSpec,
+		evalCtx.Codec,
+		tableDesc,
+		sourceIndex,
+		[]descpb.ColumnID{vectorColID},
+	); err != nil {
+		return err
+	}
+
+	vectorCol, err := catalog.MustFindColumnByID(tableDesc, vectorColID)
+	if err != nil {
+		return err
+	}
+
+	var neededColOrdinals intsets.Fast
+	neededColOrdinals.Add(vectorCol.Ordinal())
+	splitter := span.MakeSplitter(tableDesc, tableDesc.GetPrimaryIndex(), neededColOrdinals)
+	spec.FamilyIDs = splitter.FamilyIDs()
+
+	// We need to decode the columns for the PK from the vector index. If the
+	// vector index is being mutated, there's a chance the primary key is changing.
+	// The vector index needs to point into the NEW primary key, so find that here.
+	var primaryKey catalog.Index
+	primaryKey = tableDesc.GetPrimaryIndex()
+	if indexDesc.IsMutation() {
+		for _, idx := range tableDesc.NonPrimaryIndexes() {
+			if idx.GetEncodingType() == catenumpb.PrimaryIndexEncoding && !idx.IsTemporaryIndexForBackfill() {
+				primaryKey = idx
+				break
+			}
+		}
+	}
+	keyCols := primaryKey.CollectKeyColumnIDs().Ordered()
+	return rowenc.InitIndexFetchSpec(&spec.ExtractPKFetchSpec, evalCtx.Codec, tableDesc, indexDesc, keyCols)
+}
+
+// PKDecoder is used to extract the primary key from a vector index.
+type pkDecoder struct {
+	fetchSpec *fetchpb.IndexFetchSpec
+	output    rowenc.EncDatumRow
+	scratch   rowenc.EncDatumRow
+	colOrdMap catalog.TableColMap
+}
+
+// Init initializes the PKDecoder with the given fetch spec from the
+// GetFullVectorsFetchSpec.
+func (d *pkDecoder) Init(fetchSpec *fetchpb.IndexFetchSpec) {
+	d.fetchSpec = fetchSpec
+	if cap(d.output) < len(fetchSpec.FetchedColumns) {
+		d.output = make(rowenc.EncDatumRow, len(fetchSpec.FetchedColumns))
+	} else {
+		d.output = d.output[:0]
+	}
+	for i, col := range fetchSpec.FetchedColumns {
+		d.colOrdMap.Set(col.ColumnID, i)
+	}
+}
+
+// Decode decodes the primary key from the given key bytes. The key returned
+// remains valid until the next row is decoded.
+func (d *pkDecoder) Decode(
+	treeKey cspann.TreeKey, keyBytes []byte,
+) (row rowenc.EncDatumRow, err error) {
+	decodeFromKey := func(keyCols []fetchpb.IndexFetchSpec_KeyColumn, keyBytes []byte) error {
```

Comment:
nit: Feels like we could pull out a common helper from here and the vector search operator - but it's fine if that's just a TODO for now.